### PR TITLE
ReduceLROnPlateau Callback and CSVLogger Callback

### PR DESCRIPTION
**Pull request by basveeling**
I'm proposing to add two callbacks to keras. If there's interest I'll clean up according to the guidelines, otherwise no hard feelings :-). 

ReduceLROnPlateau:
Monitors a loss and reduces the learning rate with a given factor.

CSVLogger:
Writes logs to a CSV.

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -532,3 +534,108 @@ def on_epoch_end(self, epoch, logs={}):
...
         self.writer.flush()
+
+
+class ReduceLROnPlateau(Callback):
+    """
+    Reduce learning rate when a metric has stopped improving.
+    """
+
+    def __init__(self, monitor='val_loss', reduce_factor=0.1, patience=0, verbose=0, mode='auto', epsilon=1e-4,
+                 cooldown=0, min_lr=0):
```
Comment (fchollet) [inline keras/callbacks.py]: The class docstring should explain the meaning of the arguments.

Comment (basveeling) [inline keras/callbacks.py]: Done.

**Comment by basveeling** [comment]
Thanks for the feedback! I added the documentation and cleaned up the code a bit. Looking forward to hear your thoughts.

**Comment by basveeling** [comment]
Added a test that checks if the LR is successfully reduced or not depending on the parameters.

File: keras/callbacks.py [hunk 2]
```diff
@@ -532,3 +534,155 @@ def on_epoch_end(self, epoch, logs={}):
...
+        rewrite: True: overwrite existing file, False: append if
+            file exists (useful for continuing training).
+    '''
+
+    def __init__(self, filename, separator=',', rewrite=True):
+        self.sep = separator
+        self.filename = filename
+        self.rewrite = rewrite
+        self.writer = None
+        self.keys_ = None
```
Comment (fchollet) [inline keras/callbacks.py]: Any reason why you are using `keys_` instead of `keys`? In any case, the standard naming convention for private variables would be `_keys`.

File: keras/callbacks.py [hunk 3]
```diff
@@ -532,3 +534,155 @@ def on_epoch_end(self, epoch, logs={}):
...
+        ```
+
+    Arguments
+        filename: filename of the csv file, e.g. 'run/log.csv'.
+        separator: string used to separate elements in the csv file.
+        rewrite: True: overwrite existing file, False: append if
+            file exists (useful for continuing training).
+    '''
+
+    def __init__(self, filename, separator=',', rewrite=True):
```
Comment (fchollet) [inline keras/callbacks.py]: Maybe `append` would be preferable to `rewrite`?

File: keras/callbacks.py [hunk 4]
```diff
@@ -532,3 +534,155 @@ def on_epoch_end(self, epoch, logs={}):
...
+            mode, the direction is automatically inferred
+            from the name of the monitored quantity.
+        epsilon: threshold for measuring the new optimum,
+            to only focus on significant changes.
+        cooldown: number of epochs to wait before resuming
+            normal operation after lr has been reduced.
+        min_lr: lower bound on the learning rate.
+    '''
+
+    def __init__(self, monitor='val_loss', reduce_factor=0.1, patience=10,
```
Comment (fchollet) [inline keras/callbacks.py]: Maybe `factor` would be more concise than `reduce_factor`?

File: keras/callbacks.py [hunk 5]
```diff
@@ -532,3 +534,155 @@ def on_epoch_end(self, epoch, logs={}):
             summary_value.tag = name
             self.writer.add_summary(summary, epoch)
         self.writer.flush()
+
+
+class ReduceLROnPlateau(Callback):
+    '''Reduce learning rate when a metric has stopped improving.
+
+    Models often benefit from reducing the lr by a factor of
```
Comment (fchollet) [inline keras/callbacks.py]: Prefer `learning rate` to `lr`

Comment (basveeling) [inline keras/callbacks.py]: Done.

File: keras/callbacks.py (hunk 4 above)
Comment (basveeling) [inline keras/callbacks.py]: Agreed.

File: keras/callbacks.py (hunk 3 above)
Comment (basveeling) [inline keras/callbacks.py]: Agreed.

File: keras/callbacks.py (hunk 2 above)
Comment (basveeling) [inline keras/callbacks.py]: I think I did this because `keys()` is such a common function of dictionaries and such, while it is an attribute in this case. On second thought it should be fine given the context.

**Comment by basveeling** [comment]
Accepted all suggested changes and rebased on top of master. Thanks for the careful consideration!

**Comment by fchollet** [comment]
Test failing: `E       assert (0.009999999776482582 - 0.01) > 1e-07`

Might want to reduce sensitivity and seed the RNG.

File: keras/callbacks.py [hunk 6]
```diff
@@ -534,6 +536,161 @@ def on_epoch_end(self, epoch, logs={}):
...
+        self.monitor = monitor
+        if factor >= 1.0:
+            raise ValueError('ReduceLROnPlateau does not support a factor >= 1.0.')
+        self.factor = factor
+        self.min_lr = min_lr
+        self.epsilon = epsilon
+        self.patience = patience
+        self.verbose = verbose
+        self.cooldown = cooldown
+        self.cooldown_counter = 0  # Cooldown counter.
```
Comment (fchollet) [inline keras/callbacks.py]: No need to restate the variable name in a comment.

**Comment by fchollet** [comment]
LGTM
